src/processor/operator/intersect/intersect.cpp

```diff
@@ -160,17 +150,60 @@ void Intersect::populatePayloads(
     }
 }
 
+bool Intersect::hasNextTuplesToIntersect() {
+    tupleIdxPerBuildSide[carryBuildSideIdx]++;
+    if (tupleIdxPerBuildSide[carryBuildSideIdx] == probedFlatTuples[carryBuildSideIdx].size()) {
+        if (carryBuildSideIdx == 0) {
+            return false;
+        }
+        tupleIdxPerBuildSide[carryBuildSideIdx] = 0;
+        carryBuildSideIdx--;
+        if (!hasNextTuplesToIntersect()) {
+            return false;
+        }
+        carryBuildSideIdx++;
+    }
+    return true;
+}
+
 bool Intersect::getNextTuplesInternal() {
     do {
-        if (!children[0]->getNextTuple()) {
-            return false;
+        while (carryBuildSideIdx == -1u) {
+            if (!children[0]->getNextTuple()) {
+                return false;
+            }
+            // For each build side, probe its HT and return a vector of matched flat tuples.
+            probeHTs();
+            auto maxNumTuplesToIntersect = 1u;
+            for (auto i = 0u; i < sharedHTs.size(); i++) {
+                maxNumTuplesToIntersect *= probedFlatTuples[i].size();
+            }
+            if (maxNumTuplesToIntersect == 0) {
+                // Skip if any build side has no matches.
+                continue;
+            }
+            carryBuildSideIdx = sharedHTs.size() - 1;
+            std::fill(tupleIdxPerBuildSide.begin(), tupleIdxPerBuildSide.end(), 0);
         }
-        auto tuples = probeHTs(getProbeKeys());
-        auto listsToIntersect = fetchListsToIntersectFromTuples(tuples, isIntersectListAFlatValue);
+        // Cartesian product of all flat tuples probed from all build sides.
+        // Notice: when there are large adjacency lists in the build side, which means the list is
+        // too large to fit in a single ValueVector, we end up chunking the list as multiple tuples
+        // in FTable. Thus, when performing the intersection, we need to perform cartesian product
+        // between all flat tuples probed from all build sides.
+        auto numBuildSides = sharedHTs.size();
```

Comment:
you are using sharedHTs.size() in multiple places. Better to create an inline function

Reply:
Done.